[Bug]: `Bad Request` after partial tool call failures with Github Copilot

### Pre-submission checklist

- [x] I have read the [documentation](https://codecompanion.olimorris.dev)
- [x] I have updated the plugin and all dependencies to the latest versions
- [x] I have searched for existing [issues](https://github.com/olimorris/codecompanion.nvim/issues) and [discussions](https://github.com/olimorris/codecompanion.nvim/discussions)
- [x] My issue is not a minor or cosmetic quirk (e.g., formatting, spacing, or other non-functional details)

### Neovim version (nvim -v)

v0.12.0-dev-768+gf576b59a09

### Operating system/version

Linux

### Adapter and model

Github Copilot with gpt-4.1

### Describe the bug

When Copilot tries to do multiple tool calls in one response but some of them fail, it may stuck in `Bad Request` errors for all subsequent requests.

My guess is that Copilot expects results of ALL tool calls, whether succeeded, failed or being skipped. But CodeCompanion would stop tool execution when any of them failed or rejected, and not generate results for the skipped tool calls.

### Steps to reproduce

1. `nvim --clean -u minimal.lua`
2. `:CodeCompanionChat`
3. Enter:
    ```
    @{files}
    Read the following files and make a summary.
    * `/tmp/xxx/sample/error1.txt`
    * `/tmp/xxx/sample/error2.txt`
    * `/tmp/xxx/sample/error3.txt`
    ```
4. Wait for the file read error.
5. Enter: `Ignore this error.`
6. Copilot responses `Bad Request`, and we are stuck here unless we manually fix the chat with the debug panel.
7. `gd`, remove the tool calls that were skipped.
8. The chat can continue now.

### Expected behavior

We should be able to continue this chat, either let the LLM ignore the errors or instruct it revise its tool call requests.

CodeCompanion should either keep running other tool calls, or generate dummy results for them to let the LLM know these tool calls were skipped.

### Screenshots or recordings (optional)

```lua
{ {
    content = "You are an AI programming assistant named \"CodeCompanion\". You are currently plugged into the Neovim text editor on a user's machine.\n\nYour core tasks include:\n- Answering general programming questions.\n- Explaining how the code in a Neovim buffer works.\n- Reviewing the selected code from a Neovim buffer.\n- Generating unit tests for the selected code.\n- Proposing fixes for problems in the selected code.\n- Scaffolding code for a new workspace.\n- Finding relevant code to the user's query.\n- Proposing fixes for test failures.\n- Answering questions about Neovim.\n- Running tools.\n\nYou must:\n- Follow the user's requirements carefully and to the letter.\n- Use the context and attachments the user provides.\n- Keep your answers short and impersonal, especially if the user's context is outside your core tasks.\n- Minimize additional prose unless clarification is needed.\n- Use Markdown formatting in your answers.\n- Include the programming language name at the start of each Markdown code block.\n- Do not include line numbers in code blocks.\n- Avoid wrapping the whole response in triple backticks.\n- Only return code that's directly relevant to the task at hand. You may omit code that isn’t necessary for the solution.\n- Avoid using H1, H2 or H3 headers in your responses as these are reserved for the user.\n- Use actual line breaks in your responses; only use \"\\n\" when you want a literal backslash followed by 'n'.\n- All non-code text responses must be written in the English language indicated.\n- Multiple, different tools can be called as part of the same response.\n\nWhen given a task:\n1. Think step-by-step and, unless the user requests otherwise or the task is very simple, describe your plan in detailed pseudocode.\n2. Output the final code in a single code block, ensuring that only relevant code is included.\n3. End your response with a short suggestion for the next user turn that directly supports continuing the conversation.\n4. Provide exactly one complete reply per conversation turn.\n5. If necessary, execute multiple tools in a single turn.",
    cycle = 1,
    id = 126646444,
    opts = {
      tag = "from_config",
      visible = false
    },
    role = "system"
  }, {
    content = "create_file, file_search, get_changed_files, grep_search, insert_edit_into_file, read_file\nRead the following files and make a summary.\n\n* `/tmp/xxx/sample/error1.txt`\n* `/tmp/xxx/sample/error2.txt`\n* `/tmp/xxx/sample/error3.txt`",
    cycle = 1,
    id = 575788808,
    opts = {
      visible = true
    },
    role = "user"
  }, {
    content = "<editFileInstructions>\nBefore editing a file, ensure you have its content via the provided context or read_file tool.\nUse the insert_edit_into_file tool to modify files.\nNEVER show the code edits to the user - only call the tool. The system will apply and display the edits.\nFor each file, give a short description of what needs to be edited, then use the insert_edit_into_file tools. You can use the tool multiple times in a response, and you can keep writing text after using a tool.\nThe insert_edit_into_file tool is very smart and can understand how to apply your edits to the user's files, you just need to follow the patch format instructions carefully and to the letter.\n\n## Patch Format\n*** Begin Patch\n[PATCH]\n*** End Patch\n\nThe `[PATCH]` is the series of diffs to be applied for each edit in the file. Each diff should be in this format:\n\n [3 lines of pre-context]\n-[old code]\n+[new code]\n [3 lines of post-context]\n\nThe context blocks are 3 lines of existing code, immediately before and after the modified lines of code.\nLines to be modified should be prefixed with a `+` or `-` sign.\nUnmodified lines used in context should begin with an empty space ` `.\n\nFor example, to add a subtract method to a calculator class in Python:\n\n*** Begin Patch\n def add(self, value):\n     self.result += value\n     return self.result\n\n+def subtract(self, value):\n+    self.result -= value\n+    return self.result\n+\n def multiply(self, value):\n     self.result *= value\n     return self.result\n*** End Patch\n\nMultiple blocks of diffs should be separated by an empty line and `@@[identifier]` as detailed below.\nThe immediately preceding and after context lines are enough to locate the lines to edit. DO NOT USE line numbers anywhere in the patch.\nYou can use `@@[identifier]` to define a larger context in case the immediately before and after context is not sufficient to locate the edits. Example:\n\n@@class BaseClass(models.Model):\n [3 lines of pre-context]\n-\tpass\n+\traise NotImplementedError()\n [3 lines of post-context]\n\nYou can also use multiple `@@[identifiers]` to provide the right context if a single `@@` is not sufficient.\nExample with multiple blocks of edits and `@@` identifiers:\n\n*** Begin Patch\n@@class BaseClass(models.Model):\n@@\tdef search():\n-\t\tpass\n+\t\traise NotImplementedError()\n\n@@class Subclass(BaseClass):\n@@\tdef search():\n-\t\tpass\n+\t\traise NotImplementedError()\n*** End Patch\n\nThis format is similar to the `git diff` format; the difference is that `@@[identifiers]` uses the unique line identifiers from the preceding code instead of line numbers. We don't use line numbers anywhere since the before and after context, and `@@` identifiers are enough to locate the edits.\nIMPORTANT: Be mindful that the user may have shared attachments that contain line numbers, but these should NEVER be used in your patch. Always use the contextual format described above.The system uses fuzzy matching and confidence scoring so focus on providing enough context to uniquely identify the location.\n</editFileInstructions>",
    cycle = 1,
    id = 391413238,
    opts = {
      reference = "<tool>insert_edit_into_file</tool>",
      tag = "tool",
      visible = false
    },
    role = "system"
  }, {
    cycle = 1,
    id = 197288128,
    opts = {
      visible = false
    },
    role = "llm",
    tool_calls = { {
        _index = 0,
        ["function"] = {
          arguments = '{"filepath": "/tmp/xxx/sample/error1.txt", "start_line_number_base_zero": 0, "end_line_number_base_zero": -1}',
          name = "read_file"
        },
        id = "call_qAqGfdCJct7EXIRFwf3q9Dsy",
        type = "function"
      }, {
        _index = 1,
        ["function"] = {
          arguments = '{"filepath": "/tmp/xxx/sample/error2.txt", "start_line_number_base_zero": 0, "end_line_number_base_zero": -1}',
          name = "read_file"
        },
        id = "call_Y8rqGdpvtiCeSithQ6uOQHPI",
        type = "function"
      }, {
        _index = 2,
        ["function"] = {
          arguments = '{"filepath": "/tmp/xxx/sample/error3.txt", "start_line_number_base_zero": 0, "end_line_number_base_zero": -1}',
          name = "read_file"
        },
        id = "call_ZwRXKo1htySBl6603p7kScVL",
        type = "function"
      } }
  }, {
    content = "Error reading `/tmp/tmp/xxx/sample/error1.txt`, does not exist",
    cycle = 1,
    id = 1730352165,
    opts = {
      visible = true
    },
    role = "tool",
    tool_call_id = "call_qAqGfdCJct7EXIRFwf3q9Dsy"
  }, {
    content = "Ignore this error.",
    cycle = 2,
    id = 2036909938,
    opts = {
      visible = true
    },
    role = "user"
  } }
```

### minimal.lua file

```lua
https://github.com/olimorris/codecompanion.nvim/blob/164e95c6ff0becd6d152e06be56ccd98f53bc5a8/minimal.lua
```

### Log output (optional)

```log
[DEBUG] 2025-07-16 11:04:24
[Tools] strategies.chat.agents.tools.create_file identified
[DEBUG] 2025-07-16 11:04:24
[Tools] strategies.chat.agents.tools.file_search identified
[DEBUG] 2025-07-16 11:04:24
[Tools] strategies.chat.agents.tools.get_changed_files identified
[DEBUG] 2025-07-16 11:04:24
[Tools] strategies.chat.agents.tools.grep_search identified
[DEBUG] 2025-07-16 11:04:24
[Tools] strategies.chat.agents.tools.insert_edit_into_file identified
[DEBUG] 2025-07-16 11:04:24
[Tools] strategies.chat.agents.tools.read_file identified
[INFO] 2025-07-16 11:04:24
Chat request started
[DEBUG] 2025-07-16 11:04:24
Authorizing GitHub Copilot token
[INFO] 2025-07-16 11:04:26
Request body file: /data/tmp/nvim.rwoEg3/1.json
[DEBUG] 2025-07-16 11:04:26
Request:
{ "-sSL", "-D", "/tmp/plenary_curl_f4574a9c.headers", "-X", "POST", "-H", "Content-Type: application/json", "-H", "Copilot-Integration-Id: vscode-chat", "-H", "Editor-Version: Neovim/0.12.0", "-H", "Authorization: ****", "-d", "@/data/tmp/nvim.rwoEg3/1.json", "--retry", "3", "--retry-delay", "1", "--keepalive-time", "60", "--connect-timeout", "10", "--tcp-nodelay", "--no-buffer", "https://api.individual.githubcopilot.com/chat/completions" }
[DEBUG] 2025-07-16 11:04:29
[Tools] strategies.chat.agents.tools.read_file identified
[DEBUG] 2025-07-16 11:04:29
[Tools] strategies.chat.agents.tools.read_file identified
[DEBUG] 2025-07-16 11:04:29
[Tools] strategies.chat.agents.tools.read_file identified
[INFO] 2025-07-16 11:04:29
[Agent] Initiated
[DEBUG] 2025-07-16 11:04:29
Executor:execute - `read_file` tool
[DEBUG] 2025-07-16 11:04:29
FuncExecutor:orchestrate 1
[DEBUG] 2025-07-16 11:04:29
Args: {
  end_line_number_base_zero = -1,
  filepath = "/tmp/xxx/sample/error1.txt",
  start_line_number_base_zero = 0
}
[DEBUG] 2025-07-16 11:04:29
FuncExecutor:run
[DEBUG] 2025-07-16 11:04:29
Executor:error
[WARN] 2025-07-16 11:04:30
Tool read_file: Error reading `/tmp/tmp/xxx/sample/error1.txt`, does not exist
[DEBUG] 2025-07-16 11:04:30
[Read File Tool] Error output: { "Error reading `/tmp/tmp/xxx/sample/error1.txt`, does not exist" }
[DEBUG] 2025-07-16 11:04:30
Tool output: {
  _index = 0,
  ["function"] = {
    arguments = '{"filepath": "/tmp/xxx/sample/error1.txt", "start_line_number_base_zero": 0, "end_line_number_base_zero": -1}',
    name = "read_file"
  },
  id = "call_qAqGfdCJct7EXIRFwf3q9Dsy",
  type = "function"
}
[DEBUG] 2025-07-16 11:04:30
Executor:close
[INFO] 2025-07-16 11:04:30
Chat request finished
[INFO] 2025-07-16 11:04:30
[Agent] Completed
[INFO] 2025-07-16 11:04:42
Chat request started
[INFO] 2025-07-16 11:04:42
Request body file: /data/tmp/nvim.rwoEg3/2.json
[DEBUG] 2025-07-16 11:04:42
Request:
{ "-sSL", "-D", "/tmp/plenary_curl_a3924d90.headers", "-X", "POST", "-H", "Content-Type: application/json", "-H", "Copilot-Integration-Id: vscode-chat", "-H", "Editor-Version: Neovim/0.12.0", "-H", "Authorization: ****", "-d", "@/data/tmp/nvim.rwoEg3/2.json", "--retry", "3", "--retry-delay", "1", "--keepalive-time", "60", "--connect-timeout", "10", "--tcp-nodelay", "--no-buffer", "https://api.individual.githubcopilot.com/chat/completions" }
[ERROR] 2025-07-16 11:04:44
Error: Bad Request
```

### Minimal reproduction confirmation

- [x] Yes, I have tested and provided a `minimal.lua` file that reproduces the issue

Ah yes, that yields:

```
[ERROR] 2025-07-16 10:28:20
Error: {"error":{"message":"An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'. The following tool_call_ids did not have response messages: call_lVpK5KIed7jk3scV0MIBjsOz, call_dx9DOH3aWeS8fu8nty1edHs3","code":"invalid_request_body"}}
```

These tool calls should still be executed either way. I'll try and take a look at this later on this month. THanks for raising

--- a/doc/extending/tools.md
+++ b/doc/extending/tools.md
@@ -590,7 +590,7 @@ output = {
 
 This will notify the user with the message: `Perform the calculation 100 multiply 50?`. The user can choose to proceed, reject or cancel. The latter will cancel any tools from running.
 
-You can also customize the output if a user rejects the approval:
+You can also customize the output if a user rejects the approval or cancels the tool execution:
 
 ```lua
 output = {
@@ -603,6 +603,15 @@ output = {
   ---@return nil
   rejected = function(self, agent, cmd)
     agent.chat:add_tool_output(self, "The user declined to run the calculator tool")
+  end,
+
+  ---Cancellation message back to the LLM
+  ---@param self CodeCompanion.Tool.Calculator
+  ---@param agent CodeCompanion.Agent
+  ---@param cmd table
+  ---@return nil
+  cancelled = function(self, agent, cmd)
+    agent.chat:add_tool_output(self, "The user cancelled the execution of the calculator tool")
   end,
 },
 ```

--- a/lua/codecompanion/config.lua
+++ b/lua/codecompanion/config.lua
@@ -164,6 +164,7 @@ local defaults = {
           folds = {
             enabled = true, -- Fold tool output in the buffer?
             failure_words = { -- Words that indicate an error in the tool output. Used to apply failure highlighting
+              "cancelled",
               "error",
               "failed",
               "invalid",

--- a/lua/codecompanion/strategies/chat/agents/executor/cmd.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/cmd.lua
@@ -84,12 +84,14 @@ function CmdExecutor:run(cmd, index)
               return self.executor:setup()
             end
           else
-            return self.executor:error(cmd, string.format("Failed with code %s", code))
+            self.executor:error(cmd, string.format("Failed with code %s", code))
+            return self.executor:close()
           end
         end)
 
         if not ok then
-          return self.executor:error(cmd, string.format("Error whilst running command %s: %s", cmd, output))
+          self.executor:error(cmd, string.format("Error whilst running command %s: %s", cmd, output))
+          return self.executor:close()
         end
       end)
     end,

--- a/lua/codecompanion/strategies/chat/agents/executor/func.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/func.lua
@@ -76,7 +76,8 @@ function FuncExecutor:run(func, action, input, callback)
     end
     tool_finished = true
     if msg.status == self.executor.agent.constants.STATUS_ERROR then
-      return self.executor:error(action, msg.data or "An error occurred")
+      self.executor:error(action, msg.data or "An error occurred")
+      return self.executor:close()
     end
 
     self.executor:success(action, msg.data)
@@ -90,7 +91,8 @@ function FuncExecutor:run(func, action, input, callback)
     return func(self.executor.agent, action, input, output_handler)
   end)
   if not ok then
-    return self.executor:error(action, output)
+    self.executor:error(action, output)
+    return self.executor:close()
   end
 
   if output ~= nil then

--- a/lua/codecompanion/strategies/chat/agents/executor/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/init.lua
@@ -63,16 +63,35 @@ function Executor:setup_handlers()
     rejected = function(cmd)
       if self.tool.output and self.tool.output.rejected then
         self.tool.output.rejected(self.tool, self.agent, cmd)
+      else
+        self.agent.chat:add_tool_output(self.tool, string.format("The `%s` tool was rejected.", self.tool.name))
       end
     end,
     error = function(cmd)
       if self.tool.output and self.tool.output.error then
         self.tool.output.error(self.tool, self.agent, cmd, self.agent.stderr, self.agent.stdout or {})
+      else
+        self.agent.chat:add_tool_output(
+          self.tool,
+          string.format("There was an error when calling the `%s` tool.", self.tool.name)
+        )
+      end
+    end,
+    cancelled = function(cmd)
+      if self.tool.output and self.tool.output.cancelled then
+        self.tool.output.cancelled(self.tool, self.agent, cmd)
+      else
+        self.agent.chat:add_tool_output(self.tool, string.format("The `%s` tool was cancelled.", self.tool.name))
       end
     end,
     success = function(cmd)
       if self.tool.output and self.tool.output.success then
         self.tool.output.success(self.tool, self.agent, cmd, self.agent.stdout)
+      else
+        self.agent.chat:add_tool_output(
+          self.tool,
+          string.format("The `%s` tool has finished successfully.", self.tool.name)
+        )
       end
     end,
   }
@@ -146,9 +165,9 @@ function Executor:setup(input)
       }, function(choice)
         if not choice or choice == "Cancel" then -- No selection or cancelled
           log:debug("Executor:execute - Tool cancelled")
-          finalize_agent(self)
           self:close()
-          return
+          self.output.cancelled(cmd)
+          return self:setup()
         elseif choice == "Yes" then -- Selected yes
           log:debug("Executor:execute - Tool approved")
           self:execute(cmd, input)

--- a/lua/codecompanion/strategies/chat/agents/executor/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/init.lua
@@ -232,6 +232,7 @@ end
 ---Close the execution of the tool
 ---@return nil
 function Executor:close()
+  --TODO: This is a workaround that avoids the close method being called more than once
   if self.tool then
     log:debug("Executor:close")
     self.handlers.on_exit()

--- a/lua/codecompanion/strategies/chat/agents/executor/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/init.lua
@@ -4,6 +4,15 @@ local Queue = require("codecompanion.strategies.chat.agents.executor.queue")
 local log = require("codecompanion.utils.log")
 local util = require("codecompanion.utils")
 
+local fmt = string.format
+
+---Add a response to the chat buffer regarding a tool's execution
+---@param exec CodeCompanion.Agent.Executor
+---@msg string
+local send_response_to_chat = function(exec, msg)
+  exec.agent.chat:add_tool_output(exec.tool, msg)
+end
+
 ---@class CodeCompanion.Agent.Executor
 ---@field agent CodeCompanion.Agent
 ---@field current_cmd_tool table The current cmd tool that's being executed
@@ -64,34 +73,29 @@ function Executor:setup_handlers()
       if self.tool.output and self.tool.output.rejected then
         self.tool.output.rejected(self.tool, self.agent, cmd)
       else
-        self.agent.chat:add_tool_output(self.tool, string.format("The `%s` tool was rejected.", self.tool.name))
+        -- If no handler is set then return a default message
+        send_response_to_chat(self, fmt("User rejected `%s`", self.tool.name))
       end
     end,
     error = function(cmd)
       if self.tool.output and self.tool.output.error then
         self.tool.output.error(self.tool, self.agent, cmd, self.agent.stderr, self.agent.stdout or {})
       else
-        self.agent.chat:add_tool_output(
-          self.tool,
-          string.format("There was an error when calling the `%s` tool.", self.tool.name)
-        )
+        send_response_to_chat(self, fmt("Error calling `%s`", self.tool.name))
       end
     end,
     cancelled = function(cmd)
       if self.tool.output and self.tool.output.cancelled then
         self.tool.output.cancelled(self.tool, self.agent, cmd)
       else
-        self.agent.chat:add_tool_output(self.tool, string.format("The `%s` tool was cancelled.", self.tool.name))
+        send_response_to_chat(self, fmt("Cancelled `%s`", self.tool.name))
       end
     end,
     success = function(cmd)
       if self.tool.output and self.tool.output.success then
         self.tool.output.success(self.tool, self.agent, cmd, self.agent.stdout)
       else
-        self.agent.chat:add_tool_output(
-          self.tool,
-          string.format("The `%s` tool has finished successfully.", self.tool.name)
-        )
+        send_response_to_chat(self, fmt("Executed `%s`", self.tool.name))
       end
     end,
   }

--- a/lua/codecompanion/strategies/chat/agents/executor/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/init.lua
@@ -228,9 +228,13 @@ end
 ---Close the execution of the tool
 ---@return nil
 function Executor:close()
-  log:debug("Executor:close")
-  self.handlers.on_exit()
-  util.fire("ToolFinished", { id = self.id, name = self.tool.name, bufnr = self.agent.bufnr })
+  if self.tool then
+    log:debug("Executor:close")
+    self.handlers.on_exit()
+    util.fire("ToolFinished", { id = self.id, name = self.tool.name, bufnr = self.agent.bufnr })
+    self.tool = nil
+    self.current_cmd_tool = {}
+  end
 end
 
 return Executor

--- a/lua/codecompanion/strategies/chat/agents/executor/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/executor/init.lua
@@ -95,7 +95,7 @@ function Executor:setup(input)
   end
   if self.agent.status == self.agent.constants.STATUS_ERROR then
     log:debug("Executor:execute - Error")
-    return finalize_agent(self)
+    self:close()
   end
 
   -- Get the next tool to run
@@ -190,8 +190,7 @@ function Executor:error(action, error)
     log:warn("Tool %s: %s", self.tool.name, error)
   end
   self.output.error(action)
-  finalize_agent(self)
-  self:close()
+  self:setup()
 end
 
 ---Handle a successful completion of a tool

--- a/lua/codecompanion/strategies/chat/agents/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/init.lua
@@ -127,11 +127,6 @@ function Agent:execute(chat, tools)
   ---@param executor CodeCompanion.Agent.Executor The executor instance
   ---@param tool table The tool to run
   local function enqueue_tool(executor, tool)
-    -- If an error occurred, don't run any more tools
-    if self.status == CONSTANTS.STATUS_ERROR then
-      return
-    end
-
     local name = tool["function"].name
     local tool_config = self.tools_config[name]
     local function handle_missing_tool(tool_call, err_message)
@@ -185,7 +180,6 @@ function Agent:execute(chat, tools)
       self.tool.args = args
     end
     self.tool.opts = vim.tbl_extend("force", self.tool.opts or {}, tool_config.opts or {})
-
 
     if self.tool.env then
       local env = type(self.tool.env) == "function" and self.tool.env(vim.deepcopy(self.tool)) or {}

--- a/lua/codecompanion/strategies/chat/agents/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/init.lua
@@ -185,7 +185,6 @@ function Agent:execute(chat, tools)
       local env = type(self.tool.env) == "function" and self.tool.env(vim.deepcopy(self.tool)) or {}
       util.replace_placeholders(self.tool.cmds, env)
     end
-    self.chat:add_tool_output(self.tool, "", "")
     return executor.queue:push(self.tool)
   end
 

--- a/lua/codecompanion/strategies/chat/agents/init.lua
+++ b/lua/codecompanion/strategies/chat/agents/init.lua
@@ -186,11 +186,12 @@ function Agent:execute(chat, tools)
     end
     self.tool.opts = vim.tbl_extend("force", self.tool.opts or {}, tool_config.opts or {})
 
+
     if self.tool.env then
       local env = type(self.tool.env) == "function" and self.tool.env(vim.deepcopy(self.tool)) or {}
       util.replace_placeholders(self.tool.cmds, env)
     end
-
+    self.chat:add_tool_output(self.tool, "", "")
     return executor.queue:push(self.tool)
   end
 

--- a/lua/codecompanion/strategies/chat/init.lua
+++ b/lua/codecompanion/strategies/chat/init.lua
@@ -868,14 +868,6 @@ function Chat:submit(opts)
 
   local bufnr = self.bufnr
 
-  for tool in
-    vim.iter(self.tools):filter(function(tool)
-      return tool.function_call ~= nil and (find_tool_call(tool.function_call.id, self.messages) == nil)
-    end)
-  do
-    self:add_tool_output(tool, "", "")
-  end
-
   if opts.auto_submit then
     self.watchers:check_for_changes(self)
   else

--- a/lua/codecompanion/strategies/chat/init.lua
+++ b/lua/codecompanion/strategies/chat/init.lua
@@ -1275,7 +1275,11 @@ function Chat:add_tool_output(tool, for_llm, for_user)
   -- Ensure that tool output is merged if it has the same tool call ID
   local existing = find_tool_call(tool_call.id, self.messages)
   if existing then
-    existing.content = existing.content .. "\n\n" .. output.content
+    if existing.content ~= "" then
+      existing.content = existing.content .. "\n\n" .. output.content
+    else
+      existing.content = output.content
+    end
   else
     table.insert(self.messages, output)
   end

--- a/lua/codecompanion/strategies/chat/init.lua
+++ b/lua/codecompanion/strategies/chat/init.lua
@@ -868,6 +868,14 @@ function Chat:submit(opts)
 
   local bufnr = self.bufnr
 
+  for tool in
+    vim.iter(self.tools):filter(function(tool)
+      return tool.function_call ~= nil and (find_tool_call(tool.function_call.id, self.messages) == nil)
+    end)
+  do
+    self:add_tool_output(tool, "", "")
+  end
+
   if opts.auto_submit then
     self.watchers:check_for_changes(self)
   else

--- a/lua/codecompanion/types.lua
+++ b/lua/codecompanion/types.lua
@@ -119,6 +119,7 @@
 ---@field output.rejected? fun(self: CodeCompanion.Agent.Tool, agent: CodeCompanion.Agent, cmd: table): any Function to call if the user rejects running a command
 ---@field output.error? fun(self: CodeCompanion.Agent.Tool, agent: CodeCompanion.Agent, cmd: table, stderr: table, stdout?: table): any The function to call if an error occurs
 ---@field output.success? fun(self: CodeCompanion.Agent.Tool, agent: CodeCompanion.Agent, cmd: table, stdout: table): any Function to call if the tool is successful
+---@field output.cancelled? fun(self: CodeCompanion.Agent.Tool, agent: CodeCompanion.Agent, cmd: table): any Function to call if the tool is cancelled
 ---@field args table The arguments sent over by the LLM when making the request
 ---@field tool table The tool configuration from the config file
 

--- a/tests/config.lua
+++ b/tests/config.lua
@@ -180,6 +180,10 @@ return {
           callback = vim.fn.getcwd() .. "/tests/strategies/chat/agents/tools/stubs/cmd_queue.lua",
           description = "Cmd tool",
         },
+        ["cmd_queue_error"] = {
+          callback = vim.fn.getcwd() .. "/tests/strategies/chat/agents/tools/stubs/cmd_queue_error.lua",
+          description = "Cmd tool",
+        },
         ["mock_cmd_runner"] = {
           callback = vim.fn.getcwd() .. "/tests/strategies/chat/agents/tools/stubs/mock_cmd_runner.lua",
           description = "Cmd tool",

--- a/tests/strategies/chat/agents/executor/test_queue.lua
+++ b/tests/strategies/chat/agents/executor/test_queue.lua
@@ -68,5 +68,41 @@ T["Agent"]["queue"]["can queue functions and commands"] = function()
   -- Test that the function was called
   h.eq("Data 1 Data 2", child.lua_get([[_G._test_func]]))
 end
+T["Agent"]["queue"]["can proceed on error"] = function()
+  h.eq(vim.NIL, child.lua_get([[_G._test_order]]))
+
+  child.lua([[
+    local tools = {
+      {
+        ["function"] = {
+          arguments = { data = "Data 1" },
+          name = "func_queue",
+        },
+      },
+      {
+        ["function"] = {
+          name = "cmd_queue_error",
+        },
+      },
+      {
+        ["function"] = {
+          arguments = { data = "Data 2" },
+          name = "func_queue_2",
+        },
+      },
+    }
+    agent:execute(chat, tools)
+    vim.wait(1000)
+  ]])
+
+  -- Test order
+  h.eq(
+    "Func[Setup]->Func[Success]->Func[Exit]->Cmd[Setup]->Cmd[Error]->Cmd[Exit]->Func2[Setup]->Func2[Success]->Func2[Exit]",
+    child.lua_get([[_G._test_order]])
+  )
+
+  -- Test that the function was called
+  h.eq("Data 1 Data 2", child.lua_get([[_G._test_func]]))
+end
 
 return T

--- a/tests/strategies/chat/agents/tools/stubs/cmd_queue_error.lua
+++ b/tests/strategies/chat/agents/tools/stubs/cmd_queue_error.lua
@@ -1,5 +1,5 @@
 return {
-  name = "cmd_queue",
+  name = "cmd_queue_error",
   system_prompt = "my cmd system prompt",
   cmds = {
     { "echofanweoufqwefvcergv", "0.5" },

--- a/tests/strategies/chat/agents/tools/stubs/cmd_queue_error.lua
+++ b/tests/strategies/chat/agents/tools/stubs/cmd_queue_error.lua
@@ -1,0 +1,26 @@
+return {
+  name = "cmd_queue",
+  system_prompt = "my cmd system prompt",
+  cmds = {
+    { "echofanweoufqwefvcergv", "0.5" },
+  },
+  handlers = {
+    -- Should only be called once
+    setup = function(self)
+      _G._test_order = (_G._test_order or "") .. "->Cmd[Setup]"
+      _G._test_setup = (_G._test_setup or "") .. "Setup"
+    end,
+    -- Should only be called once
+    on_exit = function(self)
+      _G._test_order = (_G._test_order or "") .. "->Cmd[Exit]"
+      _G._test_exit = (_G._test_exit or "") .. "Exited"
+    end,
+  },
+  output = {
+    -- Should only be called once
+    error = function(self, cmd, output)
+      _G._test_order = (_G._test_order or "") .. "->Cmd[Error]"
+      _G._test_output = _G._test_output or {}
+    end,
+  },
+}

